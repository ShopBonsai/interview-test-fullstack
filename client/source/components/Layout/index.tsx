import * as React from "react";
import ProductsList from "../Products";
import Toolbar from "./Toolbar";

const Layout: React.FC = () => {
  return (
    <>
      <Toolbar />
      <ProductsList />
    </>
  );
};

export default Layout;
