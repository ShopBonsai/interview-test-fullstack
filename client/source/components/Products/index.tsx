import React from "react";
import {
  CardTitle,
  CardSubtitle,
  CardText,
  Button,
  CardBody,
  Media,
} from "reactstrap";
import "./styles.css";
import { Spinner } from "../UI/Spinner";
import { IProduct } from "../../../../interfaces/IProduct";
import { timerEnd, timerStart } from "../../utils/getServerLoad";
import { useProducts } from "../../context/products-context";
import { useCart } from "../../context/shoppingcart-context";

const ProductsList: React.FC = () => {
  const {
    filteredProducts,
    loading,
    error,
    checkStock,
    updateStock,
  } = useProducts();

  const { addToCart } = useCart();

  const onClickHandler = (id: string) => {
    if (checkStock(id)) {
      const product = updateStock(id);
      if (product) {
        addToCart(product);
      }
    }
  };

  if (loading) {
    timerStart();
    return <Spinner />;
  }

  if (error || !filteredProducts) {
    timerEnd();
    return (
      <div>
        <h3>Could not fetch data!</h3>
      </div>
    );
  }

  if (filteredProducts.length < 1) {
    timerEnd();
    return (
      <div>
        <h3>No products available</h3>
      </div>
    );
  }

  timerEnd();
  return (
    <div className="product-list">
      {filteredProducts.map((product: IProduct) => {
        const {
          color,
          description,
          quantity,
          image,
          name,
          price,
          size,
        } = product;
        return quantity > 0 ? (
          <Media key={product.id} className="product-card">
            <Media left href="#">
              <Media object src={image} alt="Product image cap" />
            </Media>
            <CardBody>
              <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
              <CardTitle>Price: {price}</CardTitle>
              <CardSubtitle>Color: {color}</CardSubtitle>
              <CardSubtitle>Size: {size}</CardSubtitle>
              <CardSubtitle>Stock: {quantity}</CardSubtitle>
              <CardText>Details: {description}</CardText>
              <Button
                color="primary"
                size="lg"
                block
                onClick={() => onClickHandler(product.id)}
              >
                Add to Cart
              </Button>
            </CardBody>
          </Media>
        ) : null;
      })}
    </div>
  );
};

export default ProductsList;
