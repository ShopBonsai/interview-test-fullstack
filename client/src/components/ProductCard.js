import React from 'react';
import {
    CardTitle,
    CardSubtitle,
    CardText,
    Button,
    CardBody,
    Media,
} from 'reactstrap';

export const ProductCard = ({ data }) => {
    const { color, description, image, name, price, quantity, size } = data;

    return (
        <Media className="product-card">
            <Media left href="#">
                <Media
                    object
                    src={image}
                    alt="Product image cap"
                    className="img-tile"
                />
            </Media>
            <CardBody>
                <CardTitle className="text-title">{name}</CardTitle>
                <CardTitle>Price: ${(price / 100).toFixed(2)}</CardTitle>
                <CardSubtitle>Color: {color}</CardSubtitle>
                <CardSubtitle>Size: {size}</CardSubtitle>
                <CardSubtitle>Quantity: {quantity}</CardSubtitle>
                <CardText>Details: {description}</CardText>
                <Button color="primary" size="lg" block>
                    Buy
                </Button>
            </CardBody>
        </Media>
    );
};
