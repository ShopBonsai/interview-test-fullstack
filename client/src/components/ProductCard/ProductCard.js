import { CardBody, CardSubtitle, CardText, CardTitle, Media } from "reactstrap";
import { ProductActions } from "../ProductActions";
import PropTypes from "prop-types";
import React from "react";

export const ProductCard = ({
  localFavourites,
  onSetFavourites,
  product,
  product: { color, description, id, image, name, price, quantity, size },
}) => {
  const isQuantityZero = Boolean(quantity === 0);

  return (
    <Media key={id} className="product-card">
      <Media left href="#">
        <Media object alt="Product image cap" src={image} />
      </Media>
      <CardBody>
        <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
        <CardTitle>Price: {price}</CardTitle>
        <CardSubtitle>Color: {color}</CardSubtitle>
        <CardSubtitle>Size: {size}</CardSubtitle>
        <CardText>Details: {description}</CardText>
        <ProductActions
          isBuyNowEnabled={!isQuantityZero}
          isNotifyEnabled={isQuantityZero}
          localFavourites={localFavourites}
          product={product}
          onSetFavourites={onSetFavourites}
        />
      </CardBody>
    </Media>
  );
};
ProductCard.propTypes = {
  localFavourites: PropTypes.arrayOf(PropTypes.string),
  onBuyNow: PropTypes.func,
  onNotifyMe: PropTypes.func,
  onSetFavourites: PropTypes.func,
  product: PropTypes.object,
};
