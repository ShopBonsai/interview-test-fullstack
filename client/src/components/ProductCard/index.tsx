import React from 'react';
import { CardTitle, CardSubtitle, CardText, Button, CardBody, Media, Label, Input, Form } from 'reactstrap';

type Props = {
  id: string;
  image: string;
  name: string;
  price: number;
  color: string;
  size: string;
  description: string;
  quantity: number;
  validateQuantity: (e: React.ChangeEvent<HTMLInputElement>, quantity) => void;
  validateNumberInput: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  addItemToCart: (e: React.FormEvent<HTMLFormElement>, item) => void;
};

const ProductCard: React.FC<Props> = (props) => {
  const {
    image,
    name,
    price,
    color,
    size,
    description,
    quantity,
    validateQuantity,
    validateNumberInput,
    addItemToCart,
  } = props;
  return (
    <Media className="product-card">
      <Media left href="#">
        <Media object src={image} alt="Product image cap" />
      </Media>
      <CardBody>
        <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
        <CardSubtitle style={{ fontSize: '1.4rem' }}>${price}</CardSubtitle>
        <CardText>Color: {color}</CardText>
        <CardText>Size: {size}</CardText>
        <CardText>Details: {description}</CardText>
        <Form data-testid="form" onSubmit={(e) => addItemToCart(e, props)}>
          <Label for="quantity">QTY:</Label>
          <Input
            id="quantity"
            aria-label="quantity-input"
            type="number"
            defaultValue={quantity < 1 ? quantity : 1}
            min="1"
            max={quantity}
            onChange={(e) => validateQuantity(e, quantity)}
            onKeyPress={validateNumberInput}
          />
          <CardText className="text-danger">{quantity} left in stock</CardText>
          <Button color="primary" size="lg" block>
            Add To Cart
          </Button>
        </Form>
      </CardBody>
    </Media>
  );
};

export default ProductCard;
