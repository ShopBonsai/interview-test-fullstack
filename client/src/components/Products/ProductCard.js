import React, { useState, useEffect } from "react";
import {
  CardTitle,
  CardSubtitle,
  CardText,
  Button,
  CardBody,
  Media,
  Popover,
  PopoverBody,
  Input,
} from "reactstrap";
import "./styles.css";
import { cartItemsVar } from "../../cache";

const ProductCard = (props) => {
  const { id, color, description, image, name, price, size } = props.product;
  const { merchant } = props;

  const [popoverOpen, setPopoverOpen] = useState(false);
  const [count, setCount] = useState(0);
  const [buttonDisabled, setButtonDisabled] = useState(false);

  const toggle = () => setPopoverOpen(!popoverOpen);
  const handleChange = (event) => {
    const newCount = parseInt(event.target.value);
    setCount(isNaN(newCount) ? 0 : newCount);
  };

  const onBuyBtnClick = () => {
    const cartItems = { ...cartItemsVar() };
    const itemCount = cartItems[id]
      ? parseInt(cartItems[id].count) + parseInt(count)
      : count;

    cartItems[id] = { id, name, image, count: itemCount };
    cartItemsVar(cartItems);

    setCount(0);

    setTimeout(() => {
      setPopoverOpen(false);
    }, 2000);
  };

  useEffect(() => {
    const newCount = parseInt(count);
    const isDisabled = isNaN(newCount) || newCount === 0;
    setButtonDisabled(isDisabled);
  }, [count]);

  return (
    <Media className="product-card">
      <Media left href="#">
        <Media object src={image} alt="Product image cap" />
      </Media>
      <CardBody>
        <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
        <CardTitle>Price: {price}</CardTitle>
        <CardSubtitle>Color: {color}</CardSubtitle>
        <CardSubtitle>Size: {size}</CardSubtitle>
        <CardSubtitle>Merchant: {merchant}</CardSubtitle>
        <CardText>Details: {description}</CardText>
        <div className="quantity-input">
          <Input
            style={{ width: 70, textAlign: "center" }}
            type="number"
            value={count}
            min="0"
            onChange={handleChange}
          ></Input>
        </div>
        <Button
          id={`button-${id}`}
          color="primary"
          size="lg"
          block
          disabled={buttonDisabled}
          onClick={onBuyBtnClick}
        >
          Buy
        </Button>
        <Popover
          placement="bottom"
          isOpen={popoverOpen}
          target={`button-${id}`}
          toggle={toggle}
        >
          <PopoverBody>Added to Cart</PopoverBody>
        </Popover>
      </CardBody>
    </Media>
  );
};

export default ProductCard;