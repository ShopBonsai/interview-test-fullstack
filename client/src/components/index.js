export { Login } from "./Login";
export { NavHeader } from "./Nav";
export { ProductsList } from "./Products";
export { Signup } from "./Signup";
