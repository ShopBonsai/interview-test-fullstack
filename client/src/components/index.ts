export { default as Products } from './Products';
export { default as Nav } from './Nav';
export { default as Cart } from './Cart';
