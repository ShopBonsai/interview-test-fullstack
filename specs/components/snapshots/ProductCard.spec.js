import React from 'react';
import renderer from 'react-test-renderer';

import { ProductCard } from '../../../client/src/components/ProductCard';
import { productData } from '../fixtures/productCard';

describe('Product Card component snapshot tests', () => {
    it('should render correctly', () => {
        const tree = renderer
            .create(<ProductCard data={productData} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
