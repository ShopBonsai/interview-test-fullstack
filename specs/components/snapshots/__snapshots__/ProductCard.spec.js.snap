// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card component snapshot tests should render correctly 1`] = `
<div
  className="product-card media"
>
  <a
    className="media-left"
    href="#"
  >
    <img
      alt="Product image cap"
      className="img-tile media-object"
      src="www.fakeurl.com/img.png"
    />
  </a>
  <div
    className="card-body"
  >
    <div
      className="text-title card-title"
    >
      product title
    </div>
    <div
      className="card-title"
    >
      Price: $
      19.00
    </div>
    <div
      className="card-subtitle"
    >
      Color: 
      #424242
    </div>
    <div
      className="card-subtitle"
    >
      Size: 
      sm
    </div>
    <div
      className="card-subtitle"
    >
      Quantity: 
      5
    </div>
    <p
      className="card-text"
    >
      Details: 
      words words words words words words words words words words 
    </p>
    <button
      aria-label={null}
      className="btn btn-primary btn-lg btn-block"
      onClick={[Function]}
    >
      Buy
    </button>
  </div>
</div>
`;
